### PR TITLE
feat(ingest/s3): type aware directory sorting

### PR DESCRIPTION
**Pull request by treff7es**
Adding comparator to dir sorting which can sort dirs as numbers if it is number

## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: metadata-ingestion/src/datahub/ingestion/source/s3/source.py [hunk 1]
```diff
@@ -182,6 +183,29 @@ def get_column_type(
...
+            if folder1.split("=", 1)[0] == folder2.split("=", 1)[0]:
+                folder1 = folder1.split("=", 1)[1]
+                folder2 = folder2.split("=", 1)[1]
+
+        num_folder1 = int(folder1)
+        num_folder2 = int(folder2)
+        if num_folder1 == num_folder2:
+            return 0
+        else:
+            return 1 if num_folder1 > num_folder2 else -1
```
Comment (asikowitz) [inline metadata-ingestion/src/datahub/ingestion/source/s3/source.py:200]: I believe this can be more concisely `num_folder1 - num_folder2`

**Review by asikowitz — APPROVED** [review]
Main question is around how to best handle folders with multiple equals signs

**Comment by treff7es** [comment]
> Main question is around how to best handle folders with multiple equals signs

Hive partitioning is usually like :
`mydir/2022/12/11`
or like this:
`mydir/year=2022/month=12/day=11` 

I think these are the most common ones and for the others it should be fine to compare as string

File: metadata-ingestion/src/datahub/ingestion/source/s3/source.py [hunk 2]
```diff
@@ -182,6 +183,29 @@ def get_column_type(
...
 
 
+def partitioned_folder_comparator(folder1: str, folder2: str) -> int:
+    # Try to convert to number and compare if the folder name is a number
+    try:
+        # Stripping = from the folder names as it most probably partition name part like year=2021
+        if "=" in folder1 and "=" in folder2:
+            if folder1.split("=", 1)[0] == folder2.split("=", 1)[0]:
+                folder1 = folder1.split("=", 1)[1]
+                folder2 = folder2.split("=", 1)[1]
```
[the remark this answers is not in the document]
Comment (treff7es) [reply metadata-ingestion/src/datahub/ingestion/source/s3/source.py:193]: Hive partitioning is usually like :
`mydir/2022/12/11`
or like this:
`mydir/year=2022/month=12/day=11`

**Pull request merged**